src/vanilla/utils/proxyMap-indexMap.ts

```diff
@@ -0,0 +1,154 @@
+import { getVersion, proxy } from '../../vanilla.ts'
+
+const maybeProxify = (x: any) => proxy({ x }).x
+
+const isProxy = (x: any) => getVersion(x) !== undefined
+
+type InternalProxyObject<K, V> = Map<K, V> & {
+  data: Array<[K, V | undefined]>
+  size: number
+  toJSON: () => Map<K, V>
+}
+
+export function proxyMap<K, V>(entries?: Iterable<[K, V]> | undefined | null) {
+  const data: Array<[K, V]> = []
+  const indexMap = new Map<K, number>()
+  const emptyIndexes: number[] = []
+
+  if (entries !== null && typeof entries !== 'undefined') {
+    if (typeof entries[Symbol.iterator] !== 'function') {
+      throw new TypeError(
+        'proxyMap:\n\tinitial state must be iterable\n\t\ttip: structure should be [[key, value]]',
+      )
+    }
+    for (const [k, v] of entries) {
+      const key = maybeProxify(k)
+      const value = maybeProxify(v)
+      indexMap.set(key, data.length)
+      data.push([key, value])
+    }
+  }
+
+  const vObject: InternalProxyObject<K, V> = {
+    data,
+    get size() {
+      return indexMap.size
+    },
+    get(key: K) {
+      const k = maybeProxify(key)
+      if (!indexMap.has(k) && !isProxy(this)) {
+        // eslint-disable-next-line @typescript-eslint/no-unused-expressions
+        this.data.length
+      }
+      if (indexMap.has(k)) {
+        const index = indexMap.get(k)!
+        if (this.data[index] !== undefined) {
+          return this.data[index]![1]
+        }
+      }
+      return undefined
+    },
+    has(key: K) {
+      const k = maybeProxify(key)
+      if (!indexMap.has(k) && !isProxy(this)) {
+        // eslint-disable-next-line @typescript-eslint/no-unused-expressions
+        this.data.length
+      }
+      return indexMap.has(k)
+    },
+    set(key: K, value: V) {
+      if (!isProxy(this)) {
+        if (import.meta.env?.MODE !== 'production') {
+          throw new Error('Cannot perform mutations on a snapshot')
+        } else {
+          return this
+        }
+      }
+      const k = maybeProxify(key)
+      const v = maybeProxify(value)
+      if (indexMap.has(k)) {
+        const index = indexMap.get(k)!
+        this.data[index] = [k, v]
+      } else {
+        const index =
+          emptyIndexes.length > 0 ? emptyIndexes.pop()! : this.data.length
+        indexMap.set(k, index)
+        this.data[index] = [k, v]
+      }
```

Comment:
micro optimization. no guarantee if it improves or not.
```suggestion
      let index = indexMap.get(k)
      if (index === undefined) {
        index = emptyIndexes.length ? emptyIndexes.pop()! : this.data.length
        indexMap.set(k, index)
      }
      this.data[index] = [k, v]
```